### ISSUE TITLE
Split binding.pyx for smaller/less memory hungry builds

### ISSUE THREAD

**Issue by touilleMan** [touilleMan/godot-python#143]
Currently all the binding code to expose the Godot classes declared by `api.json` is generated into a single `pythonscript/godot/bindings.pyx` file.

This make it very very big (~7mo for `bindings.pyx`, 87mo for the corresponding `bindings.c` generated by cython) and so very slow and memory hungry to compile (~200s, 6Go RAM with gcc, 2Go with clang). On top of that having a single file prevent parallel compilation...

The solution would be to split this big file into multiple smaller ones.

For instance having `bindings_part0.pyx`, `bindings_part1.pyx` etc. and finally a `bindings.py` that only does `from .binding_partXXX import *` to expose everything to the outer world in a clean way.
Of course given `bindings_partN.pyx` would likely requires `bindings_partN-1.pyx`, we would have to define a corresponding .pxd files (and so the final `binding.pxd` would be only be composed of `form .binding_partXXX cimport *`)

The good thing is `generate_binding.py` already order Godot classes by inheritance (https://github.com/touilleMan/godot-python/blob/72df9d1b38120aeec3c702004e14dcb82e95bde5/tools/generate_bindings.py#L387-L399), so I guess we could implement this by simply having a `max_classes_per_part` value and doing `from .bindings_partN-1 cimport *` at the beginning of each `.pxy`/`.pxd`

On top of that, the build system has to be modified (https://github.com/touilleMan/godot-python/blob/72df9d1b38120aeec3c702004e14dcb82e95bde5/SConstruct#L352-L363) with use of globbing to lazilly match the generated `binding_partXXX.pxd/pyx`

**Comment by curradium** [comment]
I agree that the ```max_classes_per_part``` value is probably the way to go; I can't think of any good way to split the bindings other than some arbitrary amount. What would be a good value for ```max_classes_per_part```? 10? 20? Or perhaps closer to 100?

(I was going to suggest this last night, but ended up thinking this would be an obvious next step for you to think of as well, so didn't mention it. :P )

**Comment by touilleMan** [comment]
I think `max_classes_per_part` should be provided as parameter to the
`generate_bindings.py` script. This way we could easily test multiple
values to find the best tradeoff between compile time and size (each .so
add it own overhead)

But if I had to guess, I would say around 70 classes per .so
so we would end up with 10 parts, which would roughly mean ~600mo RAM and
20s compilation per part, and good parallelism with modern cpu ;-)

Le ven. 31 janv. 2020 à 19:37, Russell Pagden <notifications@github.com> a
écrit :

**Comment by touilleMan** [comment]
I've worked a bit on this, see branch https://github.com/touilleMan/godot-python/tree/splitted-bindings-module

The issue is doing `cimport` on extension types imply runtime import on them.
So basically we end up with:
bindings_part0.pyx:
```cython
from bindings_part1 cimport Resource

cdef class Object:
    def foo(self, Resource arg):
        [...]
```

bindings_part1.pyx:
```cython
from bindings_part0 cimport Object

cdef class Resource(Object):
    [...]
```

This compile fine, but at runtime we have `bindings_part1.so` trying to import `bindings_part0.so` which in turn try to import `bindings_part1.so`...

The solution would be to correct `def foo(self, Resource arg)` into `def foo(self, object args)` to remove the need for the cimport.
However this mean typecheck on `arg` must be done dynamically (with a `from godot.bindings import Resource; assert isinstance(arg, Resource)`) which would be slower.
So not sure if we should do that anyway

**Comment by curradium** [comment]
Hmmmm. It seems like there's no way of getting around having some kind of compilation/runtime cost. I'd rather have it at compilation time myself, even if it does take a while + 7Go. (Or 2Go with clang!)

If I've got this straight, autopxd auto-generates Cython bindings from the godot_headers module as bindings.pxd, and then is compiled by Cython into an .so for pythonscript to call; as such, the main thing we need to optimise is the compilation/generation step.

That said, I think the best thing to do is slim down the generation of code; it's clear that the compiler is spending all that time and memory just optimising the bulky generated code. That said, upstream (autopxd) seems very absent, and I don't know how much work we could really do... and in the end, it'll all come out to the same result at the end of compilation anyway.

(P.S. Sorry for the long time since my last post!)

**Comment by touilleMan** [comment]
> autopxd auto-generates Cython bindings from the godot_headers module as bindings.pxd

In theory, yes. However autopxd output still contains errors and must be patched by hand:
https://github.com/touilleMan/godot-python/blob/035db0f1c79489cf1500f491902642f66b3fd53b/pythonscript/godot/_hazmat/gdnative_api_struct.pxd#L3-L9

Also the file generated is called `gdnative_api_struct.pxd`, `bindings.pxd` is another unrelated generated file (it is generated with jinja templates from the api.json file describing the classes exposed by Godot)

In the end  `gdnative_api_struct.pxd` is only a header so it doesn't impact compilation performances (basically all cython modules in the project import it and they all compile fast except the bindings one).
